### PR TITLE
Attempt to fix race with client disconnect

### LINKED ISSUE #25179: Flaky test: Microsoft.AspNetCore.Server.IIS.FunctionalTests.ClientDisconnectTests.ReaderThrowsCanceledException
```
[0.001s] [TestLifetime] [Information] Starting test ReaderThrowsCanceledException at 2020-08-23T22:10:02
[0.011s] [Microsoft.AspNetCore.Server.IIS.FunctionalTests.TestServer] [Debug] Method: GET, RequestUri: 'http://localhost:53157/start', Version: 1.1, Content: <null>, Headers:
{
}
[0.013s] [Microsoft.Extensions.Hosting.Internal.Host] [Debug] Hosting starting
[0.013s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Debug] Loaded hosting startup assembly IIS.Tests
[0.013s] [Microsoft.Hosting.Lifetime] [Information] Application started. Press Ctrl+C to shut down.
[0.013s] [Microsoft.Hosting.Lifetime] [Information] Hosting environment: Production
[0.013s] [Microsoft.Hosting.Lifetime] [Information] Content root path: C:\h\w\B3EB0A08\w\AD720913\e\.
[0.014s] [Microsoft.Extensions.Hosting.Internal.Host] [Debug] Hosting started
[0.014s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request starting HTTP/1.1 GET http://localhost:53157/start - -
[0.014s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request finished HTTP/1.1 GET http://localhost:53157/start - - - 200 - - 0.3809ms
[0.014s] [Microsoft.AspNetCore.Server.IIS.FunctionalTests.TestServer] [Debug] StatusCode: 200, ReasonPhrase: 'OK', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent, Headers:
{
  Server: Microsoft-IIS/10.0
  Date: Sun, 23 Aug 2020 22:10:02 GMT
  Content-Length: 4
}
[0.014s] [Microsoft.AspNetCore.Server.IIS.FunctionalTests.TestServer] [Debug] Done
[0.016s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request starting HTTP/1.1 POST http://localhost/ - 1
[0.016s] [Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer] [Debug] Connection ID "18014398511092596837" disconnecting.
[0.123s] [Microsoft.Hosting.Lifetime] [Information] Application is shutting down...
[0.530s] [Microsoft.AspNetCore.Server.IIS.Core.IISHttpServer] [Error] Connection ID "18014398511092596837", Request ID "80000066-0000-fa00-b63f-84710c7967bb": An unhandled exception was thrown by the application.
System.Runtime.InteropServices.COMException (0x80070006): The handle is invalid. (0x80070006 (E_HANDLE))
   at Microsoft.AspNetCore.Server.IIS.NativeMethods.HttpTryCancelIO(NativeSafeHandle pInProcessHandler) in /_/src/Servers/IIS/IIS/src/NativeMethods.cs:line 287
   at Microsoft.AspNetCore.Server.IIS.Core.IO.AsyncIOEngine.Complete() in /_/src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs:line 146
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContextOfT`1.ProcessRequestAsync() in /_/src/Servers/IIS/IIS/src/Core/IISHttpContextOfT.cs:line 126
[0.530s] [Microsoft.AspNetCore.Hosting.Diagnostics] [Information] Request finished HTTP/1.1 POST http://localhost/ - 1 - 200 - - 513.9849ms
[2.180s] [TestLifetime] [Information] Finished test ReaderThrowsCanceledException in 2.1788803s
```

cc @jkotalik

### PR DESCRIPTION
**Pull request by jkotalik**
Fixes https://github.com/dotnet/aspnetcore/issues/25179

I'm not a fan of locking on an internal field, but I needed the context to check if client disconnect has been called.

### CODE REVIEW

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ namespace Microsoft.AspNetCore.Server.IIS.Core
     internal partial class IISHttpContext
     {
         private long _consumedBytes;
+        internal bool _clientDisconnected;
```
Comment (Tratcher) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs:17]:
```suggestion
        internal bool ClientDisconnected { get; private set; }
```

**Review by Tratcher — APPROVED** [review]

File: src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs [hunk 2]
```diff
@@ -138,10 +138,15 @@ public void NotifyCompletion(int hr, int bytes)
...
-            lock (_contextSync)
+            lock (_context._contextLock)
             {
                 _stopped = true;
-                NativeMethods.HttpTryCancelIO(_handler);
+
+                // Should only call CancelIO if the client hasn't disconnected
+                if (!_context._clientDisconnected)
+                {
+                    NativeMethods.HttpTryCancelIO(_handler);
```
Comment (BrennanConroy) [inline src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs:148]: Who invalidates this handle?

Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs:148]: The IISHttpContext: https://github.com/dotnet/aspnetcore/blob/master/src/Servers/IIS/IIS/src/Core/IISHttpContext.cs#L674

**Review by BrennanConroy — APPROVED** [review]

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.cs [hunk 3]
```diff
@@ -324,7 +324,7 @@ private void EnsureIOInitialized()
             // If at this point request was not upgraded just start a normal IO engine
             if (AsyncIO == null)
             {
-                AsyncIO = new AsyncIOEngine(_contextLock, _requestNativeHandle);
+                AsyncIO = new AsyncIOEngine(this, _requestNativeHandle);
```
Comment (halter73) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:327]: If the `IISHttpContext` is only used for its lock object, shouldn't we just pass the lock object? That way _contextLock doesn't need to be made internal either.

Comment (jkotalik) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:327]: It's used for https://github.com/dotnet/aspnetcore/pull/25883/files#diff-d8f0d56e181f0c0743395b9c9295093bR146

**Pull request merged**